ntoskrnl/ke/dpc.c

```diff
@@ -914,28 +914,46 @@ KeRemoveQueueDpc(IN PKDPC Dpc)
 /*
  * @implemented
  */
+_IRQL_requires_max_(APC_LEVEL)
 VOID
 NTAPI
 KeFlushQueuedDpcs(VOID)
 {
-    PKPRCB CurrentPrcb = KeGetCurrentPrcb();
+    ULONG ProcessorIndex;
+    PKPRCB TargetPrcb;
+
     PAGED_CODE();
+    ASSERT(KeGetCurrentThread()->SystemAffinityActive == FALSE);
 
-    /* Check if this is an UP machine */
-    if (KeActiveProcessors == 1)
+    /* Loop all processors */
+    for (ProcessorIndex = 0; ProcessorIndex < KeNumberProcessors; ProcessorIndex++)
```

Comment:
Correct me if I'm wrong but according to MSDN:

>The KeFlushQueuedDpcs routine returns after all currently queued DPCs on all processors have executed their DPC routines to completion.
[...]
Drivers can use this routine to wait until all currently-queued DPCs are run. Note that KeFlushQueuedDpcs can take a long time to return, so drivers should not use it along any critical code paths.

Which implies that not only the DPCs have to be delivered but they must complete their deferred routine of which this function has to wait for. There is `DpcQueueDepth` from `KDPC_DATA` for every KPRCB that gets decremented each time the deferred routine is invoked. Probably you can use this member as an indicator to ensure that every DPC has really been flushed properly? Unless I'm missing an important detail here so again, I could be biased.

Reply:
Additional waiting is not required: The calling thread is running below DISPTACH_LEVEL. If a remote processor has a DPC queued (implying that processor is currently running at DISPATCH_LEVEL or above), KeSetSystemAffinityThread will cause the current thread to be scheduled on that processor by sending a DPC interrupt, which will only run, once that processor is running at below DISPATCH_LEVEL again and at this point all queued DPCs must have completed. This also works for threaded DPCs, since no other thread can interrupt them and thus the calling thread will not get scheduled on the target processor, until all threaded DPCs have finished.

Reply:
Makes sense a lot now. Thanks for the explanation.